sql/mozfun/map/mode_last_struct_retain_nulls/udf.sql

```diff
@@ -0,0 +1,251 @@
+/*
+Return the most frequent STRUCT from an array, breaking ties by latest occurrence
+(i.e., mode_last_retain_nulls over whole structs). Use to keep related fields aggregated together.
+See also: `map.mode_last`, which determines each value using `stats.mode_last`.
+*/
+CREATE OR REPLACE FUNCTION map.mode_last_struct_retain_nulls(entries ANY TYPE) AS (
+  (
+    SELECT AS STRUCT
+      s.*
+    FROM
+      (
+        SELECT
+          s,
+          COUNT(*) AS freq,
+          MAX(pos) AS last_pos
+        FROM
+          UNNEST(entries) AS s
+          WITH OFFSET pos
+        GROUP BY
+          s
+      )
+    ORDER BY
+      freq DESC,
+      last_pos DESC
+    LIMIT
+      1
+  )
+);
+
+-- Tests
+SELECT
+  -- 1) Most frequent wins (Berlin appears twice)
+  assert.equals(
```

Comment:
You should be able to use `assert.struct_equals` for all of these which will do `TO_JSON_STRING` for you and make this easier to read and edit.
```suggestion
  assert.struct_equals(
```